bugsnag-android-core/src/main/java/com/bugsnag/android/BugsnagMainJournal.kt

```diff
@@ -0,0 +1,175 @@
+package com.bugsnag.android
+
+import com.bugsnag.android.internal.DocumentPath
+import com.bugsnag.android.internal.JournaledDocument
+import java.io.File
+import java.io.IOException
+import java.lang.IllegalStateException
+import java.lang.Thread
+import kotlin.concurrent.thread
+
+/**
+ * The main journal contains the document that will be sent to the back end whenever an event occurs.
+ * All changes to the document must be made via journal commands to ensure that they are preserved
+ * in the event of a crash.
+ *
+ * The single instance to be used in Bugsnag is "instance", and startInstance() must be called before
+ * using it.
+ */
+class BugsnagMainJournal internal constructor(
+    private val logger: Logger,
+    baseDocumentPath: File,
+    initialDocument: Map<String, Any>
+) {
+    internal val journal = JournaledDocument(
+        baseDocumentPath,
+        journalType,
+        journalVersion,
+        mmapBufferSize,
+        highWater,
+        initialDocument
+    )
+
+    /**
+     * The document in its current state (with all current journal entries applied).
+     * Do NOT modify any of the document's contents directly; information will be lost if you do.
+     * Use addCommand() instead to modify the document.
+     */
+    val document: Map<String, Any> get() = journal
```

Comment:
Perhaps this could be enforced with `Collections.unmodifiableMap()`?

Reply:
Doesn't `Map` already enforce that?

There are actually two levels of threat here:
- Modifying the map itself
- Modifying a map or list inside of the map

Unfortunately there's no way to enforce the second invariant because we need to be able to modify via the journal command system.

Reply:
Good point - I've got too much Java on the brain :)